File: src/test/java/com/github/tetrisanalyzer/move/ValidMovesTest.java
package nu.tengstrand.tetrisanalyzer.move;

import nu.tengstrand.tetrisanalyzer.board.Board;
import nu.tengstrand.tetrisanalyzer.piece.Piece;
import nu.tengstrand.tetrisanalyzer.piece.PieceS;
import nu.tengstrand.tetrisanalyzer.piecemove.AllValidPieceMovesForEmptyBoard;
import nu.tengstrand.tetrisanalyzer.piecemove.PieceMove;
import nu.tengstrand.tetrisanalyzer.settings.DefaultGameSettings;
import nu.tengstrand.tetrisanalyzer.settings.GameSettings;
import org.junit.Test;

import java.util.*;

import static org.junit.Assert.assertEquals;

public class ValidMovesTest {

    @Test public void pieceMoves() {
        Board board = new Board(5,4);
        ValidMoves validMoves = new ValidMoves(board);
        PieceMove startPiece = getStartPiece(board);
        List<Move> moves = new ArrayList<Move>();

        for (PieceMove pieceMove : validMoves.getPieceMoves(startPiece)) {
            moves.add(pieceMove.getMove());
        }
        Collections.sort(moves, new Comparator<Move>() {
            public int compare(Move m1, Move m2) {
                if (m1.getX() < m2.getX()) { return -1; }
                if (m1.getX() > m2.getX()) { return 1; }
                if (m1.getRotation() < m2.getRotation()) { return -1; }
                if (m1.getRotation() < m2.getRotation()) { return 1; }
                if (m1.getY() < m2.getY()) { return -1; }
                return m1.getY() > m2.getY() ? 1 : 0;
            }
        });

        assertEquals(Arrays.asList(
                new Move(0,0, 2),
                new Move(1,0, 1),
                new Move(0,1, 2),
                new Move(1,1, 1),
                new Move(0,2, 2),
                new Move(1,2, 1),
                new Move(1,3, 1)), moves);
    }

    private PieceMove getStartPiece(Board board) {
        Piece piece = new PieceS();
        GameSettings settings = new DefaultGameSettings();
        AllValidPieceMovesForEmptyBoard allValidPieceMoves = new AllValidPieceMovesForEmptyBoard(board, settings);

        return allValidPieceMoves.startMoveForPiece(piece);
    }
}

File: langs/scala/src/test/scala/nu/tengstrand/tetrisanalyzer/move/ValidMovesTest.scala
package nu.tengstrand.tetrisanalyzer.move

import nu.tengstrand.tetrisanalyzer.BaseTest
import org.junit.Test
import nu.tengstrand.tetrisanalyzer.settings.DefaultGameSettings
import nu.tengstrand.tetrisanalyzer.board.Board
import nu.tengstrand.tetrisanalyzer.piecemove.{PieceMove, AllValidPieceMovesForEmptyBoard}
import nu.tengstrand.tetrisanalyzer.piece.PieceS

class ValidMovesTest extends BaseTest {

  @Test def pieceMoves(): Unit = {
    val board = Board(5,4)
    val validMoves = new ValidMoves(board)
    val startPiece = getStartPiece(board)
    val pieceMoves = validMoves.pieceMoves(startPiece)
    val moves = for { pieceMove <- pieceMoves } yield pieceMove.move

    moves.sortBy{m => (m.x, m.rotation, m.y)} should be (List(
      Move(0,0, 2),
      Move(1,0, 1),
      Move(0,1, 2),
      Move(1,1, 1),
      Move(0,2, 2),
      Move(1,2, 1),
      Move(1,3, 1))
    )
  }

  private def getStartPiece(board: Board): PieceMove = {
    val piece = PieceS()
    val settings = new DefaultGameSettings
    val allValidPieceMoves = new AllValidPieceMovesForEmptyBoard(board, settings)

    allValidPieceMoves.startMoveForPiece(piece)
  }
}